pkg/controller/kubedirectorcluster/types.go

```diff
@@ -88,17 +90,51 @@ const (
 	chmod +x ` + appPrepStartscript + ` &&
 	rm -rf /opt/guestconfig/appconfig.tgz`
 	appPrepConfigStatus = "/opt/guestconfig/configure.status"
-	appPrepConfigRunCmd = `rm -f /opt/guestconfig/configure.* &&
-	echo -n %s= > ` + appPrepConfigStatus + ` && 
-	nohup sh -c '` + appPrepStartscript + ` --configure 2>/opt/guestconfig/configure.stderr 1>/opt/guestconfig/configure.stdout;
-	echo -n $? >> ` + appPrepConfigStatus + `' &`
+
 	fileInjectionCommand = `mkdir -p %s && cd %s &&
 	curl -L %s -o %s`
-	appPrepConfigReconnectCmd = `echo -n %s= > ` + appPrepConfigStatus + ` &&
-	nohup sh -c '` + appPrepStartscript + ` --reconnect 2>/opt/guestconfig/configure.stderr 1>/opt/guestconfig/configure.stdout;
+
+	appPrepConfigTemplateCmd = `echo -n %s= > ` + appPrepConfigStatus + ` &&
+	nohup sh -c '` + appPrepStartscript + ` --%s 2>/opt/guestconfig/configure.stderr 1>/opt/guestconfig/configure.stdout;
 	echo -n $? >> ` + appPrepConfigStatus + `' &`
 )
 
+type ConfigArg string
+
+const (
+	Configure   ConfigArg = "configure"
+	Reconnect   ConfigArg = "reconnect"
+	PodUpgraded ConfigArg = "pod_upgraded"
+	PodReverted ConfigArg = "pod_reverted"
+)
+
+var cmdCache = make(map[ConfigArg]*string)
+
+// GetAppConfidCmd gets cached startscript command template for the specified `arg`
+// and substitutes containerId into this template.
+// The possible `arg` values currently are: Configure, Reconnect, PodUpgraded, PodReverted
+func GetAppConfigCmd(
+	containerId string,
+	arg ConfigArg,
+) string {
+
+	cachedConfigCmdTemplate := func() string {
+		result := cmdCache[arg]
+		if result == nil {
+			cmd := fmt.Sprintf(appPrepConfigTemplateCmd, "%s", arg)
+			// Clean all outputs before executing command with --run argument
+			if arg == Configure {
+				cmd = `rm -f /opt/guestconfig/configure.* && ` + cmd
+			}
+			result = &cmd
+			cmdCache[arg] = result
+		}
+		return *result
+	}
+
+	return fmt.Sprintf(cachedConfigCmdTemplate(), containerId)
+}
```

Comment:
Some good cleanup/factoring!

I don't think the cachedConfigCmdTemplate func needs to be defined though right? Seems like that code could just be executed as part of the GetAppConfigCmd function.

Reply:
Done